compiler/src/dotty/tools/dotc/ast/NavigateAST.scala

```diff
@@ -74,9 +74,23 @@ object NavigateAST {
       }
       path
     }
-    def singlePath(p: Positioned, path: List[Positioned]): List[Positioned] =
-      if (p.pos contains pos) childPath(p.productIterator, p :: path)
-      else path
+    def singlePath(p: Positioned, path: List[Positioned]): List[Positioned] = {
+      val currentHasPos =
+        p.pos.contains(pos) && (!p.pos.isZeroExtent ||
+          (path match {
+            case p1 :: _ => p1.pos.isZeroExtent || !p1.pos.contains(pos)
+            case _ => true
+          }))
+
+      if (currentHasPos) {
+        p match {
```

Comment:
Maybe, but I think it would be even better if we used something higher-level than dealing with productIterator by hand, that would take care of lazy trees, I'll leave a FIXME in the code.